design-proposals/shadow-node.md

```diff
@@ -0,0 +1,343 @@
+  |-----------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+  | object.spec == oldObject.spec                                                                                                     | virt-handler user cannot modify spec of the nodes                              |                                                                                                                                                                        |
+  | object.status == oldObject.status                                                                                                 | virt-handler user cannot modify status of the nodes                            |                                                                                                                                                                        |
+  | object.(not labels/annotations/resourceVersion/managedFields) == oldObject.(not labels/annotations/resourceVersion/managedFields) | virt-handler user can only change allowed sub-metadata fields.                 |                                                                                                                                                                        |
+  | size(`newNonKubevirtLabels`) == size(`oldNonKubevirtLabels`)                                                                      | virt-handler user cannot add/delete non kubevirt-owned labels                  |                                                                                                                                                                        |
+  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |
+
+### Backwards compatibility:
+validationAdmissionPolicy [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it should be OK for deployment on Kubevirt main and next minor release branches.
+However, since kubevirt needs to be supported up to 2 k8s versions back, then the deployment of these new objects needs to be conditioned by a new feature gate "node restriction". HCO main and v1.3 branches will enable this feature gate by default.
```

Comment:
Can we program the component that deploys Kubevirt (virt-operator?) to condition what and how to setup the cluster based on its version/capability?
Clusters that can support the new policy (e.g. in v1.30 or in v1.29 where the K8S FG is enabled), will have Kubevirt deployed in one way and in other clusters in the existing manner.
The condition will be dropped once we drop support for K8S 1.29.

Having a FG protecting this change (or any other) will drag this for at least 2 version (6 months), requiring users to act on making their cluster more secure. I hope we can take that burden on ourself by just giving them the best possible option given the setup.

If this is not possible, could you please elaborate on why not?

Reply:
I'm not sure it's possible, will need to check. 
But theoretically speaking, isn't that problematic to introduce such a dependency to a k8s version? It sounds unusual..

Reply:
After thinking it through, I agree with you.
a FG is not a good fit for this, since this is not something we want a user (=cluster-admin) to need to switch on, as this is not a user facing feature. 
We have 2 alternatives then:
- add a logic on runtime to check k8s version and fail only if the vesion is above 0.30.0
- always deploy and ignore on fail.

I prefer the former. Will fix the proposal accordingly